torch/distributed/_composable/fsdp/_fsdp_param_group.py

```diff
@@ -166,13 +234,46 @@ def _to_unsharded(self):
                 fsdp_param.to_unsharded()
             self._sharded_state = ShardedState.UNSHARDED
 
+    # Hook Registration #
+    def _register_post_backward_hook(
+        self, args: Tuple[Any, ...], kwargs: Dict[str, Any]
+    ) -> Tuple[Tuple[Any, ...], Dict[str, Any]]:
+        if not torch.is_grad_enabled():
+            return args, kwargs
+        args_list, args_spec = tree_flatten(args)
+        kwargs_list, kwargs_spec = tree_flatten(kwargs)
+        args_kwargs_list = list(args_list) + list(kwargs_list)
+        inp_tensor_indices: List[int] = []
+        inp_tensors: List[torch.Tensor] = []
+        for i, obj in enumerate(args_kwargs_list):
+            if not torch.is_tensor(obj) or not obj.requires_grad:
+                continue
+            inp_tensor_indices.append(i)
+            inp_tensors.append(obj)
+        if len(inp_tensors) == 0:
+            return args, kwargs  # no tensors that require gradients
+        inp_tensors = RegisterPostBackwardHook.apply(self, *inp_tensors)
+        for inp_tensor_idx, inp_tensor in zip(inp_tensor_indices, inp_tensors):
+            args_kwargs_list[inp_tensor_idx] = inp_tensor
+        args_list = args_kwargs_list[: len(args_list)]
+        kwargs_list = args_kwargs_list[len(args_list) :]
+        args = tree_unflatten(args_list, args_spec)
+        kwargs = tree_unflatten(kwargs_list, kwargs_spec)
+        return args, kwargs
+
     # Properties #
     @property
     def _all_gather_process_group(self) -> dist.ProcessGroup:
         mesh_info = self.mesh_info
         assert isinstance(mesh_info, FSDPMeshInfo)
         return mesh_info.shard_process_group
 
+    @property
+    def _reduce_scatter_process_group(self) -> dist.ProcessGroup:
```

Comment:
Good question! In the next PR, when we add `reshard_after_forward`, the all-gather process group can be a smaller PG if we resharded to that smaller PG after forward. (Because I was pulling code from the mega PR with `reshard_after_forward` implemented already, I directly added this as 2 separate properties.)